File: peasant_multiplication.swift
// http://programmingpraxis.com/2014/12/23/ancient-algorithms/

func product(left:Int,right:Int) -> Int {
  var prod:Int = 0
  var leftVal:Int = left
  var rightVal:Int = right
  while leftVal > 0 {
    if leftVal % 2 == 1 {
      prod = rightVal + prod
    }
    leftVal = leftVal / 2
    rightVal = rightVal * 2
  }
  return prod
}

println(product(3,4))

File: peasant_multiplication.pl
#!/usr/bin/perl
use strict;
use warnings;

sub product {
  my $left = shift;
  my $right = shift;
  my $result = 0;
  while ($left > 0) {
    $result += $right if ($left % 2 != 0);
    $left /= 2; 
    $right *= 2;
  }
  return $result;
}

my $val = product(3,9);
print "$val\n";
